src/transformers/models/llama/modeling_llama.py

```diff
@@ -92,61 +93,90 @@ def forward(self, hidden_states):
 
 
 class LlamaRotaryEmbedding(nn.Module):
-    def __init__(self, dim, max_position_embeddings=2048, base=10000, device=None, scaling_factor=1.0):
+    def __init__(self, dim, max_position_embeddings=2048, base=10000, device=None, scaling_factor=1.0, **kwargs):
         super().__init__()
-        self.scaling_factor = scaling_factor
-        self.dim = dim
-        self.max_position_embeddings = max_position_embeddings
-        self.base = base
-        inv_freq = 1.0 / (self.base ** (torch.arange(0, self.dim, 2, dtype=torch.int64).float().to(device) / self.dim))
+        self.rope_config = {
+            "base": base,
+            "dim": dim,
+            "max_position_embeddings": max_position_embeddings,
+            "scaling_factor": scaling_factor,
+        }
+        self.rope_config.update(kwargs)
+        # BC: in absence of "rope_type" in kwargs, set it to "default"
+        self.rope_config["rope_type"] = self.rope_config.get("rope_type", "default")
+
+        inv_freq = compute_frequencies(self.rope_config, device)
         self.register_buffer("inv_freq", inv_freq, persistent=False)
-        # For BC we register cos and sin cached
         self.max_seq_len_cached = max_position_embeddings
 
+        # BC: dynamic NTK and yarn used `scaling_factor` as a frequency parameter, not for position_ids scaling
+        if "dynamic" in self.rope_config["rope_type"] or "yarn" in self.rope_config["rope_type"]:
+            self.rope_config["scaling_factor"] = 1.0
+        # Special case: on yarn, `attention_factor` has a default suggested by the paper
+        if "yarn" in self.rope_config["rope_type"]:
+            self.rope_config["attention_factor"] = self.rope_config.get(
+                "attention_factor", 0.1 * math.log(scaling_factor) + 1.0
+            )
+
+    def dynamic_frequency_update(self, position_ids, device):
+        """dynamic RoPE layers need to recompute `inv_freq` when going beyond the original maximum sequence length"""
+        seq_len = torch.max(position_ids) + 1
+        if seq_len > self.max_seq_len_cached:
+            inv_freq = compute_frequencies(self.rope_config | {"seq_len": seq_len}, device)
+            self.register_buffer("inv_freq", inv_freq, persistent=False)  # TODO joao: may break with compilation
+            self.max_seq_len_cached = seq_len
```

Comment:
we are doing this with each forward, but if we go back to smaller seq_len, we also need to put back our old frequency